### ISSUE TITLE
Cannot specify generic class-string on templated object

### ISSUE THREAD

**Issue by michnovka** [vimeo/psalm#7913]
https://psalm.dev/r/d706b6cb58

I would expect to be able to specify that both templated version of the class `Foo` (`Foo<T>`) as well as non-templated version `Foo` can be used, but this throws `[MixedArgumentTypeCoercion](https://psalm.dev/194)`

**Comment by AndrolGenhald** [comment]
This is the same issue I mentioned [here](https://github.com/doctrine/orm/issues/9681#issuecomment-1110242687). It should be possible to fix, but it will likely be pretty difficult to implement.

If `$foo` is a `class-string<Foo>` then constructing `new $foo()` should create a `Foo`. Clearly then, if `$foo` is a `class-string<Foo<int>>`, then constructing `new $foo(...)` should create a `Foo<int>`, but (assuming `Foo::__construct` takes one argument which is the template) in this case `new $foo("bar")` creates a `Foo<string>`, which is incorrect.

The only way I can see to satisfy the requirements is to have some way to tie the template from the `class-string` to the type used in the constructor call, so that `new $foo("bar")` shows an `InvalidArgument` issue because `"bar"` is not an `int`, but this is not currently possible.

**Comment by orklah** [comment]
@AndrolGenhald can you confirm this works as expected now?

**Comment by AndrolGenhald** [comment]
@orklah Nope, this is going to require a good bit of work. Here's a reproducer: https://psalm.dev/r/78f69052ab
